repose-aggregator/services/service-client/impl/src/main/java/org/openrepose/core/services/serviceclient/akka/impl/AkkaServiceClientImpl.java

```diff
@@ -71,26 +74,49 @@ public AkkaServiceClientImpl(HttpClientService httpClientService) {
         quickFutureCache = CacheBuilder.newBuilder()
                 .expireAfterWrite(FUTURE_CACHE_TTL, FUTURE_CACHE_UNIT)
                 .build();
+    }
+
+    @PreDestroy
+    public void destroy() {
+        // tie this actor system to our bean lifecycle
+        actorSystem.shutdown();
+    }
 
+    @Override
+    public void setConnectionPoolId(String connectionPoolId) {
+        if (StringUtils.equals(this.connectionPoolId, connectionPoolId)) {
+            return;
+        }
 
-        tokenActorRef = actorSystem.actorOf(new Props(new UntypedActorFactory() {
-            public UntypedActor create() {
+        ActorRef oldTokenActorRef = tokenActorRef;
+        this.connectionPoolId = connectionPoolId;
+        createTokenActor(FORCE_CREATE);
 
-                return new AuthTokenFutureActor(serviceClient);
-            }
-        }).withRouter(new RoundRobinRouter(numberOfActors)), "authRequestRouter");
+        if (oldTokenActorRef != null) {
+            actorSystem.stop(oldTokenActorRef);
+        }
     }
 
+    private synchronized void createTokenActor(ForceCreate forceCreate) {
```

Comment:
I like that idea.

Before I put in a lot of effort to dig deeper into Akka and refactor this, do you know if these can be handled:

1. I want to lazily initialization the ServiceClient instance being passed to the AuthTokenFutureActor constructor.
2. I want to be able to update the ServiceClient instance in case the configuration is updated to use a different connection pool id and update the number of actors (based on the pool size).

I may not need to do #1 if it's cheap to always use the default connection pool id and then just update the actors to use the new connection pool id once it's set for the first time.

I may not need to do #2 if we don't want to support configuration updates to the connection pool id.

I need to support at least one of those unless I can get Spring to inject an instance of AkkaServiceClient with the configured connection pool ID (which I don't see how with the current setup).